### ISSUE TITLE
Account Ledger - Include Pending items

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#923]
In GitLab by @kdjstudios on Jun 1, 2018, 10:20

Hello Team,

After recent review of current complaints and concerns about the new vs old account ledger and primarily the lack of pending items in a non confusing display. I put together the following mock up. Please take note of the following topics.

- This is just to show how each pending item would appear. I know you can not generally have both a staged fees and pending draft invoice at the same time, etc...
- The CSV and PDF buttons at the top will only download the finalized items, while the CSV and PDF at the bottom will download both finalized and pending items.
- The edit button for each type of pending item will go to appropriate page:
 - Staged Fee, will go to the staged fee page.
 - Draft Invoice, will go to the draft invoice page.
- Draft invoices will not show a quantity. 
- The help link in the note will go to a wiki page (Will create once needed)
- We should have a 'help tool tip' on the 'estimated balance' labels and amounts.
 - Will show "This amount is estimated and may not take into account taxes and other miscellaneous charges."
-  Estimated dates for each type of item:
 - Staged fees, will show the 'next billing cycle invoice date' (Same date shown in the staged fees page)
 - Draft Invoice, will show the expected 'invoice date'

![Account_Ledger_Pending](/uploads/ea7057b75ba63b9a1e87d78cdb7c6a46/Account_Ledger_Pending.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 1, 2018, 10:21

**NOTE: This is for discussion. Please do not start developing this until discussed and approved by @tim.traylor and Gary.**

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 4, 2018, 06:16

@kdjstudios Hi Kyle, Nice mockup.  Gary will want to see this to approve.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 29, 2019, 13:53

Marking for Gary review.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 31, 2019, 09:51

@amishra @smasih @pchaudhary 

This was reviewed today and we will want this to be updated.

- top download links are only finalized items.
- bottom download links are finalized and pending items. 

- payments will continue to be part of the account ledger, regardless if they can be deleted.  
 - I will remove wording from mock up and update mock up. Once updated we will need an estimate.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:52

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:52

@amishra @smasih @pchaudhary 

Updated the mock up. Please provide an estimate.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:55

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 5, 2019, 03:54

@kdjstudios 

Thank you for the update. 

It's a great mockup. 
However before giving an estimate please provide the mock-up of the edit button.

* if we want to edit the same page. (Need Mockup)**or**
* Should we redirect to account staged fee page? (Not need).

thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 5, 2019, 10:59

@amishra 

As specified in the main ticket

> The edit button for each type of pending item will go to appropriate page:
> 
> - Staged Fee, will go to the staged fee page.
> 
> - Draft Invoice, will go to the draft invoice page.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 5, 2019, 21:56

@kdjstudios 

Thank you for the update.

Estimate:

|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|R-Small|2|Aware of Concepts, implementation tricky|3.5|Low Impact|1.0|4.5|

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 13, 2019, 14:00

Approving this. marking for development.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 2, 2019, 04:26

@kdjstudios 

Thank you for the update.

we have completed the development and moving on our testing. We will let you know when ready for your testing.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 4, 2019, 02:54

@kdjstudios 


We have complete the functionality. and only left is the calculation of estimate balance.
![image](/uploads/ba1ffc2815edca9abfb62326cdbdbeca/image.png)
I am a little bit confused about `Estimated balance column`. How to calculate estimate balance in above attachment.

Please guide us.

Thank.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 4, 2019, 09:45

@amishra 

The estimated balance will calculate just like the balance column of the finalized items. We just have to label it 'estimated', because a user can change pending items. So we do not want the users to think that those are a finalized balance and then be alarmed when it changes.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Mar 5, 2019, 07:21

@kdjstudios 

We work on this today & most of the part is developed but now we have one doubt which should be discussed with Sumeet. 

We'll get back to you in this after the discussion with him.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 5, 2019, 23:39

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 7, 2019, 24:15

@kdjstudios @tim.traylor 

We have a question?

Suppose there are two pending staged fees

User creates a draft 

staged fees are transferred to draft invoice 

and no longer has the pending flag

so 

![image](/uploads/8db52d3186c302c56709727795847a57/image.png)

**draft invoice will replace all the pending items**

Is this acceptable or do we need to build extra functionality??

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 7, 2019, 09:03

This was reviewed today. The mock as stated in the first post is only to show how each item will display. We know that only pending items will display in the pending section and that staged fees and draft invoices generally wont display at the same time.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 8, 2019, 01:06

@kdjstudios 

I would advise to move the new sub ledger in the tab like we have here. Because the ledger could be longer for the old accounts.

![image](/uploads/b13fc33321ca83aa7ccbba72060df3e0/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 8, 2019, 03:49

@kdjstudios 


![image](/uploads/a6dd73942bee9030f55468909dd435d4/image.png)
![image](/uploads/766d73426223613600673a84dcdae8de/image.png)
In the above attachments, We just want to confirm in draft invoice:
* Amount column (new charges of  the current invoice)
* Subtotal column(a little bit confused what amount we need to show. Should we same as amount column or amount?)
* Estimated column(Total amount of the current invoice.)


Please suggest me what should be shown in the subtotal column.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 8, 2019, 08:34

@pchaudhary 

Please keep everything in the same tab. Please keep in mind we already have a separate ticket opened that will allow the user to display the newer items at the top, by reversing the display order. #1374 

@amishra 

The amount and Subtotal columns will both display the "new charges" of the invoice. The estimated column will display the updated account balance; which should be the total column on the invoice.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 11, 2019, 06:14

@kdjstudios 

Thank you for the update.

We have completed the development and moving our testing. We will let you know when the branch is live for testing. 

Thank.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 12, 2019, 08:46

@amishra @smasih @pchaudhary 

Thank you, please post this on either the Merge or Staging development servers so Answernet may test this. I know we can not use the Merge2 as that is where the SD development is occurring and Dev we are waiting for response from a client on testing as well.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 11, 2019, 06:35

@kdjstudios 

We have setup branch on the dev server. Please test it.

### Test cases and results
* When account has pending staged fees.

![image](/uploads/db3c61ddcb0fd935a965fbcae60927ac/image.png)

* when we click on the edit button then redirect to account staged fee page.

* When account has draft invoices.
![image](/uploads/d60373160b5e3b0c50d0bc542608f1fc/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 11, 2019, 12:03

@amishra 

It looks like the bottom download links are not including the pending and finalized items as requested in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/923#note_34157 ; but rather are only including the pending items.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 12, 2019, 05:19

@kdjstudios 

Thank you for the update.

> It looks like the bottom download links are not including the pending and finalized items as requested in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/923#note_34157 ; but rather are only including the pending items.

We have completed the development.

Please test it again on the dev server.

> The help link in the note will go to a wiki page (Will create once needed)

Please provide me the verbiage of the help note or create the page so that we can add the link on the help URL.

Thank.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 12, 2019, 09:39

@amishra @tim.traylor 

**Multiple Download Buttons**

We now have multiple download buttons showing in the pending section. the ones in the top right of the table are correct, the ones located under the table need to be removed. 

![image](/uploads/25d7e02e44349096c606a98eb584292f/image.png)


**Help Wiki**

Please create the wiki with the following information:

Subject: Account Ledger

Body: The Account ledger will display a historical view of all transactions of payments and invoices on the account. The account ledger also includes a pending section. This pending section will contain any staged fees that have not been invoiced and also display any invoices that are currently in a draft mode and have yet to be finalized. This pending section will aid users in seeing what the estimated new account balance will be after these pending items are finalized.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 12, 2019, 23:36

@kdjstudios 

Thank you for the update.



> We now have multiple download buttons showing in the pending section. the ones in the top right of the table are correct, the ones located under the table need to be removed.

We have removed the buttons.



> Please create the wiki with the following information:

We have created the wiki help.

Please test it on the dev server.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 15, 2019, 08:23

@amishra 

I retested and got the following error when trying to view the help. The CSV and PDF download are now working. 

![image](/uploads/15fb02d2ed2f9885c2088be3877542bb/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 15, 2019, 23:07

@kdjstudios 

Thank you for the update.

We had to change the database for another testing and forgot to add the wiki that's why you get the error.

We have added the wiki so please test it again.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 16, 2019, 08:00

@amishra 

Thank you. This appears to be working as expected now. Marking for announcement.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Apr 27, 2019, 05:30

Hi [Kyle] ( @kdjstudios ), 

We have successfully released this on both of the production servers.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 27, 2019, 12:17

I was able to confirm that the pending items are now displaying in a separate section and that the exports are working.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 30, 2019, 14:48

closed
